File: Task/Equal-prime-and-composite-sums/Lua/equal-prime-and-composite-sums.lua
do --[[ find n and m where the sums of the first n primes and first m
        composites where the sums are equal
   --]]

    local function isPrime( n )         -- returns TRUE if n is prime
         if   n <= 2 then return n == 2
         elseif n % 2 == 0 then return false
         else local prime, i, r = true, 1, math.floor( math.sqrt( n ) )
              repeat
                  i = i + 2
                  if i <= r then
                       prime = n % i ~= 0
                  end
              until i > r or not prime
              return prime
         end
    end

    local  count, n, m, sumP, sumC, numP, numC = 0, 2, 1, 5, 4, 3, 4
    io.write( "           sum    primes composites\n" )

    repeat
        if sumC > sumP then
            repeat numP = numP + 2 until isPrime( numP )
            sumP = sumP + numP
            n    = n + 1
        end
        if sumP > sumC then
            repeat numC = numC + 1 until not isPrime( numC )
            sumC = sumC + numC
            m    = m + 1
        end
        if sumP == sumC then
            io.write( string.format( "%14d", sumP ), string.format( "%10d", n ), string.format( "%11d", m ), "\n" )
            count = count + 1;
            if count < 8 then
                repeat numC = numC + 1 until not isPrime( numC )
                sumC = sumC + numC
                m    = m + 1
            end
        end
    until count >= 8
end

File: Task/Equal-prime-and-composite-sums/Python/equal-prime-and-composite-sums.py
# equal_prime_comp_sums.py by Xing216
import math
import numpy
def prime_composites(upto=50000):
    nums = numpy.arange(2,upto+1)
    primes=numpy.arange(3,upto+1,2)
    isprime=numpy.ones((upto-1)//2,dtype=bool)
    for factor in primes[:int(math.sqrt(upto))//2]:
        if isprime[(factor-2)//2]: isprime[(factor*3-2)//2::factor]=0
    primes = numpy.insert(primes[isprime],0,2)
    intersect = nums[numpy.in1d(nums, primes)]
    mask1 = numpy.searchsorted(nums,intersect)
    composites = numpy.delete(nums,mask1)
    return primes, composites
primes, composites = prime_composites()
cum_primes = numpy.cumsum(primes)
cum_composites = numpy.cumsum(composites)
print("Sum        | Prime Index | Composite Index")
print("------------------------------------------")
for idx, num in enumerate(cum_primes):
    if num in cum_composites:
        print(f"{num:10,} | {idx+1:11,} | {numpy.where(cum_composites == num)[0][0]+1:15,}")
